File: Baek Joon/BOJ_1074.cpp
#include <iostream>
#include <cmath>
using namespace std;

int ans = 0;

int segmentation(int N, int r, int c) {
	if (r == 0 && c == 0)
		return 0;
	else {
		if (r < pow(2, N - 1) && c < pow(2, N - 1)) {
			ans+=segmentation(N - 1, r, c);
			return 0;
		}
		else if (r < pow(2, N - 1) && c < pow(2, N)) {
			ans+=segmentation(N - 1, r, c - pow(2, N - 1));
			return pow(4, N - 1);
		}
		else if (r < pow(2, N) && c < pow(2, N - 1)) {
			ans+=segmentation(N - 1, r - pow(2, N - 1), c);
			return pow(4, N - 1) * 2;
		}
		else {
			ans+=segmentation(N - 1, r - pow(2, N - 1), c - pow(2, N - 1));
			return pow(4, N - 1) * 3;
		}
	}
}


int main() {
	std::ios::sync_with_stdio(false);
	int N, r, c;
	cin >> N >> r >> c;

	int total=ans+segmentation(N, r, c);
	cout << total << endl;
		  
}

File: Baek Joon/Java/BOJ_1074.java
package com.ssafy.algo;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ_1074 {
	public static int N, r, c;

	public static int segmentation(int n, int row, int col) {
		if (n == 0)
			return 0;

		if (r < row + (int) Math.pow(2, n - 1) && c < col + (int) Math.pow(2, n - 1))
			return segmentation(n - 1, row, col);
		
		else if (r < row + (int) Math.pow(2, n - 1) && c < col + (int) Math.pow(2, n))
			return (int) Math.pow(4, n - 1) + segmentation(n - 1, row, col + (int) Math.pow(2, n - 1));
		
		else if (r < row + (int) Math.pow(2, n) && c < col + (int) Math.pow(2, n - 1))
			return (int) Math.pow(4, n - 1) * 2 + segmentation(n - 1, row + (int) Math.pow(2, n - 1), col);
		
		else
			return (int) Math.pow(4, n - 1) * 3
					+ segmentation(n - 1, row + (int) Math.pow(2, n - 1), col + (int) Math.pow(2, n - 1));

	}

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		r = Integer.parseInt(st.nextToken());
		c = Integer.parseInt(st.nextToken());

		int ans = segmentation(N, 0, 0);

		System.out.println(ans);

	}
}
